frontend/src/components/auth/CallBack.tsx

```diff
@@ -1,51 +1,59 @@
 import React, { useEffect } from 'react'
-// import React, { useEffect, useState } from 'react'
 import { useMutation } from '@apollo/client'
-// import { AUTHORIZE_WITH_GH, ADD_SERVICE } from '../../graphql/mutations'
-import { AUTHORIZE_WITH_GH } from '../../graphql/mutations'
+import { AUTHORIZE_WITH_GH, ADD_SERVICE } from '../../graphql/mutations'
 
 const CallBack = () => {
-  // const [servicedata, setServiceData] = useState(null)
-
   const queryString = window.location.search
   const params = new URLSearchParams(queryString)
 
   const [
     authenticate,
-    { loading: mutationLoading, error: mutationError },
+    { loading: authenticateLoading, error: authenticateError },
   ] = useMutation(AUTHORIZE_WITH_GH, {
     variables: { code: params.get('code') },
   })
 
-  // const [addService, { data: addServiceData,error: addServiceError }] = useMutation(ADD_SERVICE)
-
-  // flow
-  // 1. do the AUTHORIZE_WITH_GH mutation => returns service = {serviceName, username, email, reposurl} and token
-  // 2. call ADD_SERVICE mutation to add the service to current user
-  // 3. set localstorage token to token from AUTHORIZE_WITH_GH (contains local user id, username and githubtoken)
-
-  // also somehow disable github button when current user is not logged in locally
+  const [
+    addService,
+    {
+      data: addServiceData,
+      loading: addServiceLoading,
+      error: addServiceError,
+    },
+  ] = useMutation(ADD_SERVICE)
 
   useEffect(() => {
     ;(async () => {
       try {
-        const response: any = await authenticate()
+        const response = await authenticate()
         if (response && !response.errors) {
-          localStorage.setItem('token', response.data.authorizeWithGithub.token)
+          const token = response.data.authorizeWithGithub.token
+          const {
+            __typename,
+            ...serviceUser
+          } = response.data.authorizeWithGithub.serviceUser
```

Comment:
What is `__typename`? Also If the `AUTHORIZE_WITH_GITHUB` returns serviceUser and a token, would typing the mutations (like in `EditView`) be clearer? For example, the `REPO_STATE` query has a type `RepoStateQueryResult`. Currently everything is any

Reply:
So this is what I found out: Apollo server injects `__typename` to all returned objects that have a custom type. If I understood correctly it has something to do with caching and type unions. For example when printing out the result of the `getRepoState` query, there is a `__typename` on repoState and all files because they are custom types but branches don't have it since it is just a string array.

![typename](https://user-images.githubusercontent.com/37346883/99904860-5c69e780-2cd6-11eb-87db-f04eba812d3b.png)

The reason why this is a problem now is that we are trying to send the `serviceUser` back to the server with the `connectGitService` mutation, but GraphQL rejects it because it has an extra property (`__typename`) that is not part of the defined args. So that is why I'm separating it out of the the object. This seems to be a fairly common problem that people are having. It is actually the most commented on [feature request ](https://github.com/apollographql/apollo-feature-requests/issues/6) that hasn't been implemented yet for Apollo GraphQL.

Typing the result is also a bit problematic because we can't just pretend that the `__typename` doesn't exist on the object (that just causes the mutation to error out as explained earlier) but I've added a separate `ServiceUserReturnType` just for the return value so that the shape of the returned object is more explicit now.

Reply:
Interesting. Thanks for good explanation

Reply:
Yes thanks for explanation